### PR TITLE
Add extended wildcard support for sotw

### PR DESCRIPTION
**Pull request by valerian-roche**
This PR is implementing the new wildcard flow for sotw-xds
In order to limit the implementer complexity, I merge the logic of delta stream state with the delta stream state logic
At this stage the per-resource versioning is still ignored to avoid further impact, but this would be a natural next step to limit the amount of resources sent to envoy when only some resources are updated. It will anyway also require properly implementing the proper behavior of partial replies for non (lds|cds)

### CODE REVIEW

**Comment by valerian-roche** [comment]
This is passing tests but no unit tests added yet

File: pkg/cache/v3/cache.go [hunk 1]
```diff
@@ -122,6 +122,9 @@ type RawResponse struct {
 	// Resources to be included in the response.
 	Resources []types.ResourceWithTTL
 
+	// NextVersionMap consists of updated version mappings after this response is applied
```
Comment (atollena) [inline pkg/cache/v3/cache.go]: Comment is a bit unclear to me, as it doesn't tell what exactly is in the key and value. Suggestion:

```suggestion
	// NextVersionMap maps the resource name to the empty string for resources that were updated?
```

File: pkg/cache/v3/linear.go [hunk 2]
```diff
@@ -114,26 +114,41 @@ func NewLinearCache(typeURL string, opts ...LinearCacheOption) *LinearCache {
 }
 
 func (cache *LinearCache) respond(value chan Response, staleResources []string) {
-	var resources []types.ResourceWithTTL
-	// TODO: optimize the resources slice creations across different clients
-	if len(staleResources) == 0 {
-		resources = make([]types.ResourceWithTTL, 0, len(cache.resources))
-		for _, resource := range cache.resources {
+	resources := make([]types.ResourceWithTTL, 0, len(staleResources))
```
Comment (atollena) [inline pkg/cache/v3/linear.go:117]: aren't you losing the behavior of sending all resources when `staleResources` is empty? IIUC you still rely on it e.g. at L341 in the wildcard case.

File: pkg/cache/v3/linear.go [hunk 3]
```diff
@@ -320,34 +329,56 @@ func (cache *LinearCache) CreateWatch(request *Request, streamState stream.Strea
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
+	// If the version is not up to date, check whether any requested resource has
```
Comment (atollena) [inline pkg/cache/v3/linear.go]: why did you move this block? I intuitively would have left it at L305 to
1. make the diff smaller
2. treat `err != nil` closer to where it is returned.

File: pkg/cache/v3/linear.go [hunk 4]
```diff
@@ -320,34 +329,56 @@ func (cache *LinearCache) CreateWatch(request *Request, streamState stream.Strea
...
+	// been updated between the last version and the current version. This avoids the problem
+	// of sending empty updates whenever an irrelevant resource changes.
+	stale := false
+	var staleResources map[string]struct{} // empty means all
+
 	if err != nil {
+		// The request does not include a version or the version could not be parsed.
+		// It will send all resources matching the request with no regards to the version.
```
Comment (atollena) [inline pkg/cache/v3/linear.go]: Avoid future tense in technical documentation. https://developers.google.com/style/tense

```suggestion
		// Send all resources matching the request with no regards to the version.
```

File: pkg/cache/v3/linear.go [hunk 5]
```diff
@@ -320,34 +329,56 @@ func (cache *LinearCache) CreateWatch(request *Request, streamState stream.Strea
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
+	// If the version is not up to date, check whether any requested resource has
+	// been updated between the last version and the current version. This avoids the problem
+	// of sending empty updates whenever an irrelevant resource changes.
+	stale := false
+	var staleResources map[string]struct{} // empty means all
```
Comment (atollena) [inline pkg/cache/v3/linear.go]: why the change from `[]string{}` to `map[string]struct{}`? If I read the code right, you didn't change any access pattern, did you? It might make sense to change the signature of `GetSubscribedResourceNames` to return a slice instead of the map directly.

File: pkg/cache/v3/resource.go [hunk 6]
```diff
@@ -117,16 +117,16 @@ func MarshalResource(resource types.Resource) (types.MarshaledResource, error) {
 
 // GetResourceReferences returns a map of dependent resources keyed by resource type, given a map of resources.
 // (EDS cluster names for CDS, RDS/SRDS routes names for LDS, RDS route names for SRDS).
-func GetResourceReferences(resources map[string]types.ResourceWithTTL) map[resource.Type]map[string]bool {
```
Comment (atollena) [inline pkg/cache/v3/resource.go:120]: is this change needed? There's no definite style preference between map[string]bool and map[string]struct{}, effective go suggests the former [1], as does The Go Programming Language book.

If you want to standardize on one or the other I'd do that in a separate PR, since this doesn't affect the logic and clutters the diff.

[1] https://go.dev/doc/effective_go#maps

File: pkg/cache/v3/simple.go [hunk 7]
```diff
@@ -298,17 +298,8 @@ func (cache *snapshotCache) ClearSnapshot(node string) {
...
-	set := make(map[string]bool, len(names))
-	for _, name := range names {
-		set[name] = true
-	}
-	return set
-}
-
 // superset checks that all resources are listed in the names set.
-func superset(names map[string]bool, resources map[string]types.ResourceWithTTL) error {
+func superset(names map[string]struct{}, resources map[string]types.ResourceWithTTL) error {
```
Comment (atollena) [inline pkg/cache/v3/simple.go:302]: same comment about `map[string]bool` -> `map[string]struct{}`, you may prefer the former but I'd at least separate the diffs, and in general wouldn't bother changing that in somebody else's project since it's a matter of preference, especially if you're trying to optimize for fast reviews.

File: pkg/cache/v3/simple.go [hunk 8]
```diff
@@ -570,8 +585,20 @@ func (cache *snapshotCache) Fetch(ctx context.Context, request *Request) (Respon
...
+		state := stream.NewStreamState(len(request.ResourceNames) == 0, nil)
+		wantedResources := make(map[string]struct{}, len(request.ResourceNames))
+		for _, resourceName := range request.ResourceNames {
+			if resourceName == "*" {
+				state.SetWildcard(true)
+				continue
+			}
+			wantedResources[resourceName] = struct{}{}
+		}
+		state.SetSubscribedResourceNames(wantedResources)
```
Comment (atollena) [inline pkg/cache/v3/simple.go]: Couldn't this be extracted to a constructor, e.g.

```
// NewStreamState creates a new stream state based the set of requested resources.
func NewStreamState(resources []string) {
    ....
}
```

That way you don't have to duplicate the code in sotw.server.go.

File: pkg/server/sotw/v3/server.go [hunk 9]
```diff
@@ -183,35 +162,35 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
...
-				// If these requirements aren't satisfied, leave an open watch.
-				if w.nonce == "" || w.nonce == nonce {
-					w.close()
-
-					watches.addWatch(typeURL, &watch{
-						cancel:   s.cache.CreateWatch(req, streamState, responder),
-						response: responder,
-					})
+				if w.nonce != "" && w.nonce != nonce {
+					// The request received does not match the current state of the type, we ignore it and keep our current watch
```
Comment (atollena) [inline pkg/server/sotw/v3/server.go]:
```suggestion
					// The request received does not match the current state for the resource type, we ignore it and keep our current watch
```

File: pkg/server/sotw/v3/server.go [hunk 10]
```diff
@@ -255,3 +236,27 @@ func (s *server) StreamHandler(stream stream.Stream, typeURL string) error {
 
 	return s.process(stream, reqCh, typeURL)
 }
+
+func (s *server) registerResourceNames(resources []string, streamState *stream.StreamState) {
+	if len(resources) == 0 && streamState.IsWildcard() {
+		// The xDS protocol states that if there has never been any resource set, the request should be considered wildcard
+		// This would theoretically require keeping track on whether we ever became non-empty.
+		// As it is also technically allowed to return resources which have not been subscribed to, it is a best effort here.
```
Comment (atollena) [inline pkg/server/sotw/v3/server.go]: I don't really understand this comment... Is the scenario you're thinking of:

```
- Stream requests some resources (not considered wildcard)?
- Then unsubscribes those resources, hence ending up with no resources
In theory this should not be considered wildcard, but for us it does and it's fine since we only end up sending too unrequested resources?
```

File: pkg/server/sotw/v3/server.go [hunk 11]
```diff
@@ -255,3 +236,27 @@ func (s *server) StreamHandler(stream stream.Stream, typeURL string) error {
...
+		// We do not track '*' as a resource name to avoid confusion in further processing and rely on the IsWildcard method instead
+		if resourceName == "*" {
+			wantsWildcard = true
+			continue
+		}
+		wantedResources[resourceName] = struct{}{}
+	}
+	streamState.SetWildcard(wantsWildcard)
+	streamState.SetSubscribedResourceNames(wantedResources)
+}
```
Comment (atollena) [inline pkg/server/sotw/v3/server.go]: shouldn't this logic be encapsulated inside streamstate?

File: pkg/cache/v3/linear.go (hunk 2 above)
Comment (valerian-roche) [reply pkg/cache/v3/linear.go:117]: Not sure I'm following. `respond` is never called on wildcard streams
On non-wildcard streams we should not return anything is staleResources is empty

File: pkg/cache/v3/linear.go (hunk 3 above)
Comment (valerian-roche) [reply pkg/cache/v3/linear.go]: Moved it back. Found it made it harder to follow objects as they are declared before other unrelated logic

File: pkg/cache/v3/cache.go (hunk 1 above)
Comment (valerian-roche) [reply pkg/cache/v3/cache.go]: Updated

File: pkg/cache/v3/linear.go (hunk 4 above)
Comment (valerian-roche) [reply pkg/cache/v3/linear.go]: Updated

File: pkg/cache/v3/linear.go (hunk 5 above)
Comment (valerian-roche) [reply pkg/cache/v3/linear.go]: The main reason is that I am merging objects in-between the delta and sotw servers (which are rigorously identical outside of types and subscribe/unsubscribe
Here I aligned on delta as delta really uses the map for search. I can do a separate PR for that but given the review pace I'm not sure what would be preferred

File: pkg/cache/v3/resource.go (hunk 6 above)
Comment (valerian-roche) [reply pkg/cache/v3/resource.go:120]: Same remark as above: this change is driven by unifying delta and sotw which used different structs or worse the same structs with distinct members based on delta or sotw for the exact same thing
I don't have a strong opinion on bool vs. struct{}, though I still tend to find bool as misleading

File: pkg/cache/v3/simple.go (hunk 8 above)
Comment (valerian-roche) [reply pkg/cache/v3/simple.go]: This behavior is actually different in delta, so it's a bit complex to do
The xDS protocol created a weird exception on this behavior depending on the delta vs. sotw

Comment (valerian-roche) [reply pkg/cache/v3/simple.go]: I'll see how this can be updated, but not sure I can fully hide this complexity

File: pkg/server/sotw/v3/server.go (hunk 10 above)
Comment (valerian-roche) [reply pkg/server/sotw/v3/server.go]: The case is: 
- stream is not wildcard
- then is made wildcard
- then has no resource at all
According to xDS this should be considered as "no resources", but this requires keeping a history of whether we ever left legacy mode (i.e. did we ever have resources set in any way, including *)
As the protocol also states that it's okay to return too much, I chose to ignore this case (which should anyway basically never occur in envoy today)

File: pkg/cache/v3/cache.go [hunk 12]
```diff
@@ -122,6 +122,9 @@ type RawResponse struct {
 	// Resources to be included in the response.
 	Resources []types.ResourceWithTTL
 
+	// NextVersionMap maps the resource name to the empty string for resources that were included in the response.
```
Comment (atollena) [inline pkg/cache/v3/cache.go:125]: But it's not always the empty string, right? Sometimes it's the version number, like at https://github.com/valerian-roche/go-control-plane/pull/4/files#diff-56212115e92b3629f0824a8e684c2ba8e1d70afc055edd1dd936aea206ccf707R457 ? How do readers differentiate those cases?

**Comment by atollena** [comment]
A few more things to make reviewing easier, beside the code:

- Use the DataDog fork rather than your personal one. That makes it more obvious that this is code we use in production at DataDog and that it was written in a professional context, potentially with help from other engineers, solving a company problem, by opposition with something you hacked on the side for your side project.

- Link to the relevant Github issue. If there is no issue, creating one would be worth it (ideally that should be done before opening the PR).
 
- Edit the PR description to make it more structured. Generally the PR description should be structured as:

  1. what problem is this PR solving?
  2. how is it solving the problem? should readers know something before diving into the code?
  3. what's the results of this PR? performance gain, bug fixed, new feature implemented.

- Here is my reaction reading the description:

> Add extended wildcard support for sotw

good.

> This PR is implementing the new wildcard flow for sotw-xds

First on the form: fix punctuation by adding periods at end of sentence, this shows that you proof-read yourself and that you value your reader's time.

Then: what is "the new wildcard flow"? Links to the relevant sections.

Suggestion:

In version 1.21.0, Envoy added support for `*` resource to make it a resource that can be subscribed to and unsubscribed
from at any time. See the Envoy 1.21.0 changelog [1] and the patch to the protocol overview [2].

This PR implements this behavior change in the state of the world xDS server (sotw). In order to limit the implementer complexity, I merge the logic of delta stream state with the delta stream state logic. At this stage the per-resource versioning is still ignored to avoid further impact, but this would be a natural next step to limit the amount of resources sent to envoy when only some resources are updated. It will anyway also require properly implementing the proper behavior of partial replies for non (lds|cds).

+ Add a description of what users can and cannot do before and after this commit.

[1] https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.21/v1.21.0#incompatible-behavior-changes
[2] https://github.com/envoyproxy/envoy/commit/b164962eb367e054df7fd563acce59663b16dd86

File: pkg/cache/v3/linear.go (hunk 2 above)
Comment (atollena) [reply pkg/cache/v3/linear.go:117]: > Not sure I'm following. respond is never called on wildcard streams

Makes sense, I did not see that when I read the code and there's no indication that respond shouldn't be called in with wildcard. Documenting `respond` might help readers to understand its purpose.

File: pkg/cache/v3/linear.go (hunk 5 above)
Comment (atollena) [reply pkg/cache/v3/linear.go]: I think review time is generally proportional to the size of PRs. The smaller the PRs, the faster the reviews. This is true in every project I see that actually does code review (and not just rubber stamping like in certain repos...).

**Comment by valerian-roche** [comment]
Now handled on upstream repo
